alphazero/mcts/mcts.py

```diff
@@ -307,7 +326,51 @@ def forward(root_node: MCTSNode):
     # TODO Maybe just iterate through children once, more efficient?
     # Currently doing it twice due to output difference,
     # just do forloop instead?
+    # This can def be optimized better
     return (
-        sum([child.action * child.n_visit for child in start_node.children]),
+        torch.pow(
+            sum(
+                [child.action * child.n_visit for child in start_node.children]
+            ),
+            (1 / temperature),
+        ),
         sampled_node,
     )
+
+
+def self_play(
+    game: GameClassType,
+    net: torch.nn.Module,
+    n_mcts_iter: int = 1600,
+    temperature: float = 1.0,
+    dirichlet_eps: float = 0.25,
+    dirichlet_conc: float = 1.0,
+) -> List[tuple]:
+    pos = game.get_initial_board()
+    data = []
+    is_terminal = game.is_game_over(pos)
+    player = 0
+    node = MCTSNode(state=pos, player=player)
+    n_players = game.n_players()
+    while not is_terminal:
+        policy, node = mcts(
+            start_node=node,
+            game=game,
+            net=net,
+            n_iter=n_mcts_iter,
+            temperature=temperature,
+            dirichlet_eps=dirichlet_eps,
+            dirichlet_conc=dirichlet_conc,
+        )
+        data.append((pos, policy, player))
+        pos = game.board_after_move(
```

Comment:
Why should we keep updating `pos`: wouldn't `node.state` give us what we want?
Ok, after thinking, it can be a good thing to keep the position of the root `pos`: when storing the position at each node, we might get memory issues when the tree becomes very large. In such case, keeping only the position of the root would be necessary. Let's wait and see.

Reply:
You're right. I need to think about your second comment.